### PR TITLE
Refactor globalConfiguration / WebProvider

### PR DESCRIPTION
**Pull request by Juliens**
- split the globalconfiguration to a new package (not in package server)
 - split traefik configuration and global configuration
 - webprovider is now in package provider
 - the run func need a global configuration and not a traefik configuration

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -203,9 +207,15 @@ func (server *Server) setupServerEntryPoint(newServerEntryPointName string, newS
...
-		serverMiddlewares = append(serverMiddlewares, statsRecorder)
+	if server.globalConfiguration.Web != nil {
+		metricsMiddleware := thoas_stats.New()
+		server.globalConfiguration.Web.MetricsRecorder = metricsMiddleware
+		serverMiddlewares = append(serverMiddlewares, metricsMiddleware)
+		if server.globalConfiguration.Web.Statistics != nil {
+			statsRecorder := middlewares.NewStatsRecorder(server.globalConfiguration.Web.Statistics.RecentErrors)
+			server.globalConfiguration.Web.StatsRecorder = statsRecorder
+			serverMiddlewares = append(serverMiddlewares, statsRecorder)
+		}
```
Comment (nmengin) [inline server/server.go]: @Juliens What is your mind about the deletion of the `statsRecorder` variable? 
```go
if server.globalConfiguration.Web.Statistics != nil {
    server.globalConfiguration.Web.StatsRecorder = middlewares.NewStatsRecorder(server.globalConfiguration.Web.Statistics.RecentErrors)
    serverMiddlewares = append(serverMiddlewares, server.globalConfiguration.Web.StatsRecorder)
    }
```

File: server/server.go [hunk 2]
```diff
@@ -203,9 +207,15 @@ func (server *Server) setupServerEntryPoint(newServerEntryPointName string, newS
...
 		serverMiddlewares = append(serverMiddlewares, middlewares.NewMetricsWrapper(metrics))
 	}
-	if server.globalConfiguration.Web != nil && server.globalConfiguration.Web.Statistics != nil {
-		statsRecorder = middlewares.NewStatsRecorder(server.globalConfiguration.Web.Statistics.RecentErrors)
-		serverMiddlewares = append(serverMiddlewares, statsRecorder)
+	if server.globalConfiguration.Web != nil {
+		metricsMiddleware := thoas_stats.New()
+		server.globalConfiguration.Web.MetricsRecorder = metricsMiddleware
+		serverMiddlewares = append(serverMiddlewares, metricsMiddleware)
+		if server.globalConfiguration.Web.Statistics != nil {
```
Comment (nmengin) [inline server/server.go]: @Juliens What is your mind about the deletion of the `metricsMiddleware` variable? 
```go
if server.globalConfiguration.Web.Statistics != nil {
    server.globalConfiguration.Web.MetricsRecorder = thoas_stats.New()
    serverMiddlewares = append(serverMiddlewares, server.globalConfiguration.Web.MetricsRecorder)
    }
```

Comment (Juliens) [inline server/server.go]: ok for me

File: server/server.go (hunk 1 above)
Comment (Juliens) [inline server/server.go]: ok for me

File: server/server.go [hunk 3]
```diff
@@ -260,16 +264,20 @@ func (server *Server) startHTTPServers() {
...
 	if server.accessLoggerMiddleware != nil {
 		serverMiddlewares = append(serverMiddlewares, server.accessLoggerMiddleware)
 	}
 	if server.metricsRegistry.IsEnabled() {
 		serverMiddlewares = append(serverMiddlewares, middlewares.NewMetricsWrapper(server.metricsRegistry, newServerEntryPointName))
 	}
-	if server.globalConfiguration.Web != nil && server.globalConfiguration.Web.Statistics != nil {
-		statsRecorder = middlewares.NewStatsRecorder(server.globalConfiguration.Web.Statistics.RecentErrors)
-		serverMiddlewares = append(serverMiddlewares, statsRecorder)
+	if server.globalConfiguration.Web != nil {
```
Comment (emilevauge) [inline server/server.go]: I wonder if these changes are compliant with the metrics refactor made in https://github.com/containous/traefik/pull/1968

Comment (ldez) [inline server/server.go]: ping @marco-jantke

**Comment by dtomcej** [comment]
This may seem like a silly question, but if we are migrating the web provider to its own middleware, can we use this opportunity to rename it to something like 'api', or 'rest' or some sort of API-like provider name?

It can be very confusing for people to have the API and the Admin panel be named "web"

**Comment by timoreimann** [comment]
+1 to what @dtomcej said.

**Comment by ldez** [comment]
It's not only a rest api but it's also an endpoint for the webui. A good name is hard to find.

**Comment by timoreimann** [comment]
True...

File: cmd/traefik/traefik.go [hunk 4]
```diff
@@ -210,23 +225,33 @@ Complete documentation is available at https://traefik.io`,
 	os.Exit(0)
 }
 
-func run(traefikConfiguration *server.TraefikConfiguration) {
+func run(globalConfiguration *configuration.GlobalConfiguration) {
 	fmtlog.SetFlags(fmtlog.Lshortfile | fmtlog.LstdFlags)
 
-	// load global configuration
-	globalConfiguration := traefikConfiguration.GlobalConfiguration
+	http.DefaultTransport.(*http.Transport).MaxIdleConnsPerHost = globalConfiguration.MaxIdleConnsPerHost
```
[the remark this answers is not in the document]
Comment (marco-jantke) [reply cmd/traefik/traefik.go]: This is actually like code before the introduction of the timeouts PR https://github.com/containous/traefik/pull/1873 and gets reintroduced here, which it shouldn't. Can you align your version please.

File: provider/web/web.go [hunk 5]
```diff
@@ -22,23 +22,20 @@ import (
...
+	KeyFile               string            `description:"SSL certificate"`
+	ReadOnly              bool              `description:"Enable read only API"`
+	Statistics            *types.Statistics `description:"Enable more detailed statistics"`
+	Metrics               *types.Metrics    `description:"Enable a metrics exporter"`
+	Path                  string            `description:"Root path for dashboard and API"`
+	Auth                  *types.Auth
+	Debug                 bool
+	CurrentConfigurations *safe.Safe
+	MetricsRecorder       *thoas_stats.Stats
+	StatsRecorder         *middlewares.StatsRecorder
```
[the remark this answers is not in the document]
Comment (marco-jantke) [reply provider/web/web.go]: Imho the naming is confusing. you rename `MetricsRecorder` with `Stats` to make the distinction with the existing `Metrics` implementation clear?

To be sure, fields like `StatsRecorder` can not be configured as part of the Web provider now?

File: server/server.go (hunk 3 above)
Comment (marco-jantke) [reply server/server.go]: AFAICS it doesn't interfere with the metrics implementation but the naming confuses at the moment. Its a good step by the way to get rid of the global `stats` variable. Anyhow `MetricsRecorder` is a confusing naming. I would stick with `Stats`.

**Comment by emilevauge** [comment]
>I have the feeling the package naming is quite verbose, as one example configuration.TraeifkConfiguration. Should we maybe use only and consistently config like config.TraefikConfig?

Hey @marco-jantke, we all use automatic completion today, so I think we can live we "long" package names ;)

**Pull request merged**